{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/saket/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88\n",
      "  return f(*args, **kwds)\n",
      "/home/saket/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88\n",
      "  return f(*args, **kwds)\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "import fnmatch\n",
    "import errno    \n",
    "import glob\n",
    "from pysradb import SRAdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "db = SRAdb('/home/saket/SRAmetadb.sqlite')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# List tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['metaInfo',\n",
       " 'submission',\n",
       " 'study',\n",
       " 'sample',\n",
       " 'experiment',\n",
       " 'run',\n",
       " 'sra',\n",
       " 'sra_ft',\n",
       " 'sra_ft_content',\n",
       " 'sra_ft_segments',\n",
       " 'sra_ft_segdir',\n",
       " 'col_desc',\n",
       " 'fastq']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sra_tables = db.list_tables()\n",
    "sra_tables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# List fields in a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['study_ID',\n",
       " 'study_alias',\n",
       " 'study_accession',\n",
       " 'study_title',\n",
       " 'study_type',\n",
       " 'study_abstract',\n",
       " 'broker_name',\n",
       " 'center_name',\n",
       " 'center_project_name',\n",
       " 'study_description',\n",
       " 'related_studies',\n",
       " 'primary_study',\n",
       " 'sra_link',\n",
       " 'study_url_link',\n",
       " 'xref_link',\n",
       " 'study_entrez_link',\n",
       " 'ddbj_link',\n",
       " 'ena_link',\n",
       " 'study_attribute',\n",
       " 'submission_accession',\n",
       " 'sradb_updated']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.list_fields('study')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Describe columns in a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cid</th>\n",
       "      <th>name</th>\n",
       "      <th>dtype</th>\n",
       "      <th>notnull</th>\n",
       "      <th>dflt_value</th>\n",
       "      <th>pk</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>study_ID</td>\n",
       "      <td>REAL</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>study_alias</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>study_accession</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>study_title</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>study_type</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>study_abstract</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>broker_name</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>center_name</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>center_project_name</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>study_description</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>related_studies</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>primary_study</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>sra_link</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>study_url_link</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>xref_link</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>study_entrez_link</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>ddbj_link</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>17</td>\n",
       "      <td>ena_link</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>18</td>\n",
       "      <td>study_attribute</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>19</td>\n",
       "      <td>submission_accession</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>20</td>\n",
       "      <td>sradb_updated</td>\n",
       "      <td>TEXT</td>\n",
       "      <td>0</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   cid                  name dtype notnull dflt_value pk\n",
       "0    0              study_ID  REAL       0       None  0\n",
       "1    1           study_alias  TEXT       0       None  0\n",
       "2    2       study_accession  TEXT       0       None  0\n",
       "3    3           study_title  TEXT       0       None  0\n",
       "4    4            study_type  TEXT       0       None  0\n",
       "5    5        study_abstract  TEXT       0       None  0\n",
       "6    6           broker_name  TEXT       0       None  0\n",
       "7    7           center_name  TEXT       0       None  0\n",
       "8    8   center_project_name  TEXT       0       None  0\n",
       "9    9     study_description  TEXT       0       None  0\n",
       "10  10       related_studies  TEXT       0       None  0\n",
       "11  11         primary_study  TEXT       0       None  0\n",
       "12  12              sra_link  TEXT       0       None  0\n",
       "13  13        study_url_link  TEXT       0       None  0\n",
       "14  14             xref_link  TEXT       0       None  0\n",
       "15  15     study_entrez_link  TEXT       0       None  0\n",
       "16  16             ddbj_link  TEXT       0       None  0\n",
       "17  17              ena_link  TEXT       0       None  0\n",
       "18  18       study_attribute  TEXT       0       None  0\n",
       "19  19  submission_accession  TEXT       0       None  0\n",
       "20  20         sradb_updated  TEXT       0       None  0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.desc_table('study')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SRR_bamFile',\n",
       " 'SRX_bamFile',\n",
       " 'SRX_fastqFTP',\n",
       " 'adapter_spec',\n",
       " 'anonymized_name',\n",
       " 'base_caller',\n",
       " 'bases',\n",
       " 'center_project_name',\n",
       " 'common_name',\n",
       " 'description',\n",
       " 'design_description',\n",
       " 'experiment_ID',\n",
       " 'experiment_accession',\n",
       " 'experiment_alias',\n",
       " 'experiment_attribute',\n",
       " 'experiment_entrez_link',\n",
       " 'experiment_name',\n",
       " 'experiment_title',\n",
       " 'experiment_url_link',\n",
       " 'individual_name',\n",
       " 'instrument_model',\n",
       " 'instrument_name',\n",
       " 'library_construction_protocol',\n",
       " 'library_layout',\n",
       " 'library_name',\n",
       " 'library_selection',\n",
       " 'library_source',\n",
       " 'library_strategy',\n",
       " 'multiplier',\n",
       " 'number_of_levels',\n",
       " 'platform',\n",
       " 'platform_parameters',\n",
       " 'primary_study',\n",
       " 'qtype',\n",
       " 'quality_scorer',\n",
       " 'read_spec',\n",
       " 'related_studies',\n",
       " 'run_ID',\n",
       " 'run_accession',\n",
       " 'run_alias',\n",
       " 'run_attribute',\n",
       " 'run_center',\n",
       " 'run_date',\n",
       " 'run_entrez_link',\n",
       " 'run_url_link',\n",
       " 'sample_ID',\n",
       " 'sample_accession',\n",
       " 'sample_alias',\n",
       " 'sample_attribute',\n",
       " 'sample_entrez_link',\n",
       " 'sample_name',\n",
       " 'sample_url_link',\n",
       " 'sequence_space',\n",
       " 'spots',\n",
       " 'sradb_updated',\n",
       " 'study_ID',\n",
       " 'study_abstract',\n",
       " 'study_accession',\n",
       " 'study_alias',\n",
       " 'study_attribute',\n",
       " 'study_description',\n",
       " 'study_entrez_link',\n",
       " 'study_name',\n",
       " 'study_title',\n",
       " 'study_type',\n",
       " 'study_url_link',\n",
       " 'submission_ID',\n",
       " 'submission_accession',\n",
       " 'submission_center',\n",
       " 'submission_comment',\n",
       " 'submission_date',\n",
       " 'submission_lab',\n",
       " 'taxon_id',\n",
       " 'updated_date']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted(db.desc_table('sra_ft').name.tolist())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>broker_name</th>\n",
       "      <th>center_name</th>\n",
       "      <th>center_project_name</th>\n",
       "      <th>ddbj_link</th>\n",
       "      <th>ena_link</th>\n",
       "      <th>primary_study</th>\n",
       "      <th>related_studies</th>\n",
       "      <th>sra_link</th>\n",
       "      <th>sradb_updated</th>\n",
       "      <th>study_ID</th>\n",
       "      <th>...</th>\n",
       "      <th>study_accession</th>\n",
       "      <th>study_alias</th>\n",
       "      <th>study_attribute</th>\n",
       "      <th>study_description</th>\n",
       "      <th>study_entrez_link</th>\n",
       "      <th>study_title</th>\n",
       "      <th>study_type</th>\n",
       "      <th>study_url_link</th>\n",
       "      <th>submission_accession</th>\n",
       "      <th>xref_link</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>KEIO</td>\n",
       "      <td>Bacillus subtilis subsp. natto BEST195</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2016-11-20 16:42:15</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRP000001</td>\n",
       "      <td>DRP000001</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Bacillus subtilis subsp. natto BEST195 genome ...</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000001</td>\n",
       "      <td>pubmed: 20398357 || pubmed: 25329997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>KEIO</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2017-09-21 20:23:35</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRP000002</td>\n",
       "      <td>DRP000002</td>\n",
       "      <td>None</td>\n",
       "      <td>Whole genome resequencing of B. subtilis subti...</td>\n",
       "      <td>None</td>\n",
       "      <td>Model organism for prokaryotic cell differenti...</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "      <td>None</td>\n",
       "      <td>DRA000002</td>\n",
       "      <td>pubmed: 20398357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>UT-MGS</td>\n",
       "      <td>Integrative Transcriptome Analysis</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2016-11-20 16:42:16</td>\n",
       "      <td>3.0</td>\n",
       "      <td>...</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>DRP000003</td>\n",
       "      <td>None</td>\n",
       "      <td>Although recent studies have revealed that the...</td>\n",
       "      <td>None</td>\n",
       "      <td>Comprehensive identification and characterizat...</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "      <td>DBTSS: http://dbtss.hgc.jp/</td>\n",
       "      <td>DRA000003</td>\n",
       "      <td>pubmed: 20400770</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  broker_name center_name                         center_project_name  \\\n",
       "0        None        KEIO      Bacillus subtilis subsp. natto BEST195   \n",
       "1        None        KEIO  Bacillus subtilis subsp. subtilis str. 168   \n",
       "2        None      UT-MGS          Integrative Transcriptome Analysis   \n",
       "\n",
       "  ddbj_link ena_link primary_study related_studies sra_link  \\\n",
       "0      None     None          None            None     None   \n",
       "1      None     None          None            None     None   \n",
       "2      None     None          None            None     None   \n",
       "\n",
       "         sradb_updated  study_ID                  ...                   \\\n",
       "0  2016-11-20 16:42:15       1.0                  ...                    \n",
       "1  2017-09-21 20:23:35       2.0                  ...                    \n",
       "2  2016-11-20 16:42:16       3.0                  ...                    \n",
       "\n",
       "  study_accession study_alias study_attribute  \\\n",
       "0       DRP000001   DRP000001            None   \n",
       "1       DRP000002   DRP000002            None   \n",
       "2       DRP000003   DRP000003            None   \n",
       "\n",
       "                                   study_description study_entrez_link  \\\n",
       "0                                               None              None   \n",
       "1  Whole genome resequencing of B. subtilis subti...              None   \n",
       "2  Although recent studies have revealed that the...              None   \n",
       "\n",
       "                                         study_title               study_type  \\\n",
       "0  Bacillus subtilis subsp. natto BEST195 genome ...  Whole Genome Sequencing   \n",
       "1  Model organism for prokaryotic cell differenti...  Whole Genome Sequencing   \n",
       "2  Comprehensive identification and characterizat...   Transcriptome Analysis   \n",
       "\n",
       "                study_url_link submission_accession  \\\n",
       "0                         None            DRA000001   \n",
       "1                         None            DRA000002   \n",
       "2  DBTSS: http://dbtss.hgc.jp/            DRA000003   \n",
       "\n",
       "                              xref_link  \n",
       "0  pubmed: 20398357 || pubmed: 25329997  \n",
       "1                      pubmed: 20398357  \n",
       "2                      pubmed: 20400770  \n",
       "\n",
       "[3 rows x 21 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.get_query('SELECT * FROM study LIMIT 3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Number of rows in each table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>metaInfo</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>submission</th>\n",
       "      <td>1017441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>study</th>\n",
       "      <td>161391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <td>4439622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experiment</th>\n",
       "      <td>5011580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run</th>\n",
       "      <td>5667504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra</th>\n",
       "      <td>5568559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_ft</th>\n",
       "      <td>5568559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_ft_content</th>\n",
       "      <td>5568559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_ft_segments</th>\n",
       "      <td>3007295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sra_ft_segdir</th>\n",
       "      <td>1011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>col_desc</th>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fastq</th>\n",
       "      <td>5637550</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   count\n",
       "metaInfo               2\n",
       "submission       1017441\n",
       "study             161391\n",
       "sample           4439622\n",
       "experiment       5011580\n",
       "run              5667504\n",
       "sra              5568559\n",
       "sra_ft           5568559\n",
       "sra_ft_content   5568559\n",
       "sra_ft_segments  3007295\n",
       "sra_ft_segdir       1011\n",
       "col_desc             129\n",
       "fastq            5637550"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.get_table_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Which study types are the most prominent?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number</th>\n",
       "      <th>StudyType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>45968</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>44591</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>38516</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16149</td>\n",
       "      <td>Transcriptome Analysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>14415</td>\n",
       "      <td>Metagenomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>763</td>\n",
       "      <td>Population Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>607</td>\n",
       "      <td>Epigenetics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>230</td>\n",
       "      <td>Exome Sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>110</td>\n",
       "      <td>Cancer Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>31</td>\n",
       "      <td>Pooled Clone Sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>Synthetic Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>Transcriptome Sequencing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>Whole Genome Sequencing</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Number                 StudyType\n",
       "0    45968   Whole Genome Sequencing\n",
       "1    44591                      None\n",
       "2    38516                     Other\n",
       "3    16149    Transcriptome Analysis\n",
       "4    14415              Metagenomics\n",
       "5      763       Population Genomics\n",
       "6      607               Epigenetics\n",
       "7      230          Exome Sequencing\n",
       "8      110           Cancer Genomics\n",
       "9       31   Pooled Clone Sequencing\n",
       "10       9        Synthetic Genomics\n",
       "11       1  Transcriptome Sequencing\n",
       "12       1  Whole Genome Sequencing "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.get_query('SELECT study_type AS StudyType, count(*) AS Number FROM \"study\" GROUP BY study_type order by Number DESC')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Which Sequencing Instruments are most common?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Experiments</th>\n",
       "      <th>Instrument Model</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>8704</td>\n",
       "      <td>454 GS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>986</td>\n",
       "      <td>454 GS 20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>49413</td>\n",
       "      <td>454 GS FLX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>10</td>\n",
       "      <td>454 GS FLX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>148483</td>\n",
       "      <td>454 GS FLX Titanium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16702</td>\n",
       "      <td>454 GS FLX+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>23956</td>\n",
       "      <td>454 GS Junior</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>1</td>\n",
       "      <td>AB 310 Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>111</td>\n",
       "      <td>AB 3130 Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>55</td>\n",
       "      <td>AB 3130xL Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>71</td>\n",
       "      <td>AB 3500 Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>14</td>\n",
       "      <td>AB 3500xL Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>260</td>\n",
       "      <td>AB 3730 Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1168</td>\n",
       "      <td>AB 3730xL Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>2111</td>\n",
       "      <td>AB 5500 Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>5138</td>\n",
       "      <td>AB 5500xl Genetic Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>203</td>\n",
       "      <td>AB 5500xl-W Genetic Analysis System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>323</td>\n",
       "      <td>AB SOLiD 3 Plus System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>11025</td>\n",
       "      <td>AB SOLiD 4 System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>153</td>\n",
       "      <td>AB SOLiD 4hq System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>3</td>\n",
       "      <td>AB SOLiD PI System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>501</td>\n",
       "      <td>AB SOLiD System</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>483</td>\n",
       "      <td>AB SOLiD System 2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>2581</td>\n",
       "      <td>AB SOLiD System 3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>319</td>\n",
       "      <td>BGISEQ-500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>4276</td>\n",
       "      <td>Complete Genomics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>3937</td>\n",
       "      <td>Helicos HeliScope</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>18</td>\n",
       "      <td>HiSeq X Five</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>53240</td>\n",
       "      <td>HiSeq X Ten</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>19075</td>\n",
       "      <td>Illumina Genome Analyzer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>112444</td>\n",
       "      <td>Illumina Genome Analyzer II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>63588</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>2</td>\n",
       "      <td>Illumina Genome Analyzer IIx</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3675</td>\n",
       "      <td>Illumina HiScanSQ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>14584</td>\n",
       "      <td>Illumina HiSeq 1000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>9759</td>\n",
       "      <td>Illumina HiSeq 1500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1352258</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>572648</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>6082</td>\n",
       "      <td>Illumina HiSeq 3000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>29911</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>634094</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>1</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>361</td>\n",
       "      <td>Illumina MiniSeq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>18</td>\n",
       "      <td>Illumina NextSeq 500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>41</td>\n",
       "      <td>Ion S5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>31574</td>\n",
       "      <td>Ion Torrent PGM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>8420</td>\n",
       "      <td>Ion Torrent Proton</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>857</td>\n",
       "      <td>MinION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>89496</td>\n",
       "      <td>NextSeq 500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>3259</td>\n",
       "      <td>NextSeq 550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>12056</td>\n",
       "      <td>PacBio RS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>14435</td>\n",
       "      <td>PacBio RS II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>53</td>\n",
       "      <td>Sequel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>32933</td>\n",
       "      <td>unspecified</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1665711</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Experiments                     Instrument Model\n",
       "21         8704                               454 GS\n",
       "32          986                            454 GS 20\n",
       "9         49413                           454 GS FLX\n",
       "50           10                          454 GS FLX \n",
       "4        148483                  454 GS FLX Titanium\n",
       "15        16702                          454 GS FLX+\n",
       "13        23956                        454 GS Junior\n",
       "53            1              AB 310 Genetic Analyzer\n",
       "42          111             AB 3130 Genetic Analyzer\n",
       "44           55           AB 3130xL Genetic Analyzer\n",
       "43           71             AB 3500 Genetic Analyzer\n",
       "49           14           AB 3500xL Genetic Analyzer\n",
       "39          260             AB 3730 Genetic Analyzer\n",
       "31         1168           AB 3730xL Genetic Analyzer\n",
       "30         2111             AB 5500 Genetic Analyzer\n",
       "24         5138           AB 5500xl Genetic Analyzer\n",
       "40          203  AB 5500xl-W Genetic Analysis System\n",
       "37          323               AB SOLiD 3 Plus System\n",
       "19        11025                    AB SOLiD 4 System\n",
       "41          153                  AB SOLiD 4hq System\n",
       "51            3                   AB SOLiD PI System\n",
       "34          501                      AB SOLiD System\n",
       "35          483                  AB SOLiD System 2.0\n",
       "29         2581                  AB SOLiD System 3.0\n",
       "38          319                           BGISEQ-500\n",
       "25         4276                    Complete Genomics\n",
       "26         3937                    Helicos HeliScope\n",
       "47           18                         HiSeq X Five\n",
       "8         53240                          HiSeq X Ten\n",
       "14        19075             Illumina Genome Analyzer\n",
       "5        112444          Illumina Genome Analyzer II\n",
       "7         63588         Illumina Genome Analyzer IIx\n",
       "52            2        Illumina Genome Analyzer IIx \n",
       "27         3675                    Illumina HiScanSQ\n",
       "16        14584                  Illumina HiSeq 1000\n",
       "20         9759                  Illumina HiSeq 1500\n",
       "1       1352258                  Illumina HiSeq 2000\n",
       "3        572648                  Illumina HiSeq 2500\n",
       "23         6082                  Illumina HiSeq 3000\n",
       "12        29911                  Illumina HiSeq 4000\n",
       "2        634094                       Illumina MiSeq\n",
       "54            1                      Illumina MiSeq \n",
       "36          361                     Illumina MiniSeq\n",
       "48           18                 Illumina NextSeq 500\n",
       "46           41                               Ion S5\n",
       "11        31574                      Ion Torrent PGM\n",
       "22         8420                   Ion Torrent Proton\n",
       "33          857                               MinION\n",
       "6         89496                          NextSeq 500\n",
       "28         3259                          NextSeq 550\n",
       "18        12056                            PacBio RS\n",
       "17        14435                         PacBio RS II\n",
       "45           53                               Sequel\n",
       "10        32933                          unspecified\n",
       "0       1665711                                 None"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "instruments = db.get_query('SELECT instrument_model AS \"Instrument Model\", count( * ) AS Experiments FROM \"experiment\" GROUP BY instrument_model order by Experiments DESC').sort_values(by=['Instrument Model'])\n",
    "instruments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Which assays are most common?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Library Strategy</th>\n",
       "      <th>Runs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>1665711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WGS</td>\n",
       "      <td>1092231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>690441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>627662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OTHER</td>\n",
       "      <td>314145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WXS</td>\n",
       "      <td>237358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CLONE</td>\n",
       "      <td>89555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>82832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>POOLCLONE</td>\n",
       "      <td>53738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Bisulfite-Seq</td>\n",
       "      <td>32335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SELEX</td>\n",
       "      <td>25875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>miRNA-Seq</td>\n",
       "      <td>22405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>WGA</td>\n",
       "      <td>21078</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>RAD-Seq</td>\n",
       "      <td>10290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Targeted-Capture</td>\n",
       "      <td>8427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ATAC-seq</td>\n",
       "      <td>6060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>5360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EST</td>\n",
       "      <td>3601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>RIP-Seq</td>\n",
       "      <td>3264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DNase-Hypersensitivity</td>\n",
       "      <td>2868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>MeDIP-Seq</td>\n",
       "      <td>2442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>MNase-Seq</td>\n",
       "      <td>2328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MRE-Seq</td>\n",
       "      <td>2073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>FL-cDNA</td>\n",
       "      <td>2010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Tn-Seq</td>\n",
       "      <td>1886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>WCS</td>\n",
       "      <td>1809</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>MBD-Seq</td>\n",
       "      <td>1690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CLONEEND</td>\n",
       "      <td>523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>FAIRE-seq</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Hi-C</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>CTS</td>\n",
       "      <td>282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>other</td>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Synthetic-Long-Read</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>FINISHING</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>ChIA-PET</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>VALIDATION</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Library Strategy     Runs\n",
       "0                     None  1665711\n",
       "1                      WGS  1092231\n",
       "2                 AMPLICON   690441\n",
       "3                  RNA-Seq   627662\n",
       "4                    OTHER   314145\n",
       "5                      WXS   237358\n",
       "6                    CLONE    89555\n",
       "7                 ChIP-Seq    82832\n",
       "8                POOLCLONE    53738\n",
       "9            Bisulfite-Seq    32335\n",
       "10                   SELEX    25875\n",
       "11               miRNA-Seq    22405\n",
       "12                     WGA    21078\n",
       "13                 RAD-Seq    10290\n",
       "14        Targeted-Capture     8427\n",
       "15                ATAC-seq     6060\n",
       "16               ncRNA-Seq     5360\n",
       "17                     EST     3601\n",
       "18                 RIP-Seq     3264\n",
       "19  DNase-Hypersensitivity     2868\n",
       "20               MeDIP-Seq     2442\n",
       "21               MNase-Seq     2328\n",
       "22                 MRE-Seq     2073\n",
       "23                 FL-cDNA     2010\n",
       "24                  Tn-Seq     1886\n",
       "25                     WCS     1809\n",
       "26                 MBD-Seq     1690\n",
       "27                CLONEEND      523\n",
       "28               FAIRE-seq      477\n",
       "29                    Hi-C      323\n",
       "30                     CTS      282\n",
       "31                   other      214\n",
       "32     Synthetic-Long-Read      195\n",
       "33               FINISHING       40\n",
       "34                ChIA-PET       30\n",
       "35              VALIDATION       22"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assays = db.get_query('SELECT library_strategy AS \"Library Strategy\", count( * ) AS Runs FROM \"experiment\" GROUP BY library_strategy order by Runs DESC')\n",
    "assays"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# List SRA Run table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_name</th>\n",
       "      <th>bases</th>\n",
       "      <th>spots</th>\n",
       "      <th>adapter_spec</th>\n",
       "      <th>avg_read_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217028</td>\n",
       "      <td>GSM1063575: 293T_GFP; Homo sapiens; RNA-Seq</td>\n",
       "      <td>SRR648667</td>\n",
       "      <td>9606</td>\n",
       "      <td>other</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>1806641316</td>\n",
       "      <td>50184481</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217029</td>\n",
       "      <td>GSM1063576: 293T_GFP_2hrs_severe_Heat_Shock; H...</td>\n",
       "      <td>SRR648668</td>\n",
       "      <td>9606</td>\n",
       "      <td>other</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>3436984836</td>\n",
       "      <td>95471801</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217030</td>\n",
       "      <td>GSM1063577: 293T_Hspa1a; Homo sapiens; RNA-Seq</td>\n",
       "      <td>SRR648669</td>\n",
       "      <td>9606</td>\n",
       "      <td>other</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>3330909216</td>\n",
       "      <td>92525256</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217031</td>\n",
       "      <td>GSM1063578: 293T_Hspa1a_2hrs_severe_Heat_Shock...</td>\n",
       "      <td>SRR648670</td>\n",
       "      <td>9606</td>\n",
       "      <td>other</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>3622123512</td>\n",
       "      <td>100614542</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217956</td>\n",
       "      <td>GSM794854: 3T3-Control-Riboseq; Mus musculus; ...</td>\n",
       "      <td>SRR649752</td>\n",
       "      <td>10090</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>594945396</td>\n",
       "      <td>16526261</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217957</td>\n",
       "      <td>GSM794855: 3T3-8hrs-mild-HS-Riboseq; Mus muscu...</td>\n",
       "      <td>SRR649753</td>\n",
       "      <td>10090</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>730843272</td>\n",
       "      <td>20301202</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217958</td>\n",
       "      <td>GSM794856: 3T3-2hrs-severe-HS-Riboseq; Mus mus...</td>\n",
       "      <td>SRR649754</td>\n",
       "      <td>10090</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>1785538908</td>\n",
       "      <td>49598303</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217027</td>\n",
       "      <td>GSM1063574: 3T3-Hsp70_inhibitor_3hr; Mus muscu...</td>\n",
       "      <td>SRR648666</td>\n",
       "      <td>10090</td>\n",
       "      <td>other</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>5932552920</td>\n",
       "      <td>148313823</td>\n",
       "      <td>None</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217959</td>\n",
       "      <td>GSM794857: 3T3-Control-Riboseq-rep2; Mus muscu...</td>\n",
       "      <td>SRR649755</td>\n",
       "      <td>10090</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>3713863640</td>\n",
       "      <td>92846591</td>\n",
       "      <td>None</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217960</td>\n",
       "      <td>GSM794858: 3T3-8hrs-mild-HS-Riboseq-rep2; Mus ...</td>\n",
       "      <td>SRR649756</td>\n",
       "      <td>10090</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>3967599960</td>\n",
       "      <td>99189999</td>\n",
       "      <td>None</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217961</td>\n",
       "      <td>GSM794859: 3T3-2hrs-severe-HS-Riboseq-rep2; Mu...</td>\n",
       "      <td>SRR649757</td>\n",
       "      <td>10090</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>3458211480</td>\n",
       "      <td>86455287</td>\n",
       "      <td>None</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217962</td>\n",
       "      <td>GSM794860: 3T3-8hrs-mild-2hrs-severe-HS-Ribose...</td>\n",
       "      <td>SRR649758</td>\n",
       "      <td>10090</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>3891505080</td>\n",
       "      <td>97287627</td>\n",
       "      <td>None</td>\n",
       "      <td>40.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217953</td>\n",
       "      <td>GSM794848: 3T3-Control-RNAseq; Mus musculus; R...</td>\n",
       "      <td>SRR649749</td>\n",
       "      <td>10090</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PAIRED -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>1766712168</td>\n",
       "      <td>24537669</td>\n",
       "      <td>None</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217954</td>\n",
       "      <td>GSM794849: 3T3-8hrs-mild-HS-RNAseq; Mus muscul...</td>\n",
       "      <td>SRR649750</td>\n",
       "      <td>10090</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PAIRED -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>1960403976</td>\n",
       "      <td>27227833</td>\n",
       "      <td>None</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRP017942</td>\n",
       "      <td>SRX217955</td>\n",
       "      <td>GSM794850: 3T3-2hrs-severe-HS-RNAseq; Mus musc...</td>\n",
       "      <td>SRR649751</td>\n",
       "      <td>10090</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PAIRED -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>None</td>\n",
       "      <td>1393136640</td>\n",
       "      <td>19349120</td>\n",
       "      <td>None</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   study_accession experiment_accession  \\\n",
       "0        SRP017942            SRX217028   \n",
       "1        SRP017942            SRX217029   \n",
       "2        SRP017942            SRX217030   \n",
       "3        SRP017942            SRX217031   \n",
       "4        SRP017942            SRX217956   \n",
       "5        SRP017942            SRX217957   \n",
       "6        SRP017942            SRX217958   \n",
       "7        SRP017942            SRX217027   \n",
       "8        SRP017942            SRX217959   \n",
       "9        SRP017942            SRX217960   \n",
       "10       SRP017942            SRX217961   \n",
       "11       SRP017942            SRX217962   \n",
       "12       SRP017942            SRX217953   \n",
       "13       SRP017942            SRX217954   \n",
       "14       SRP017942            SRX217955   \n",
       "\n",
       "                                     experiment_title run_accession  taxon_id  \\\n",
       "0         GSM1063575: 293T_GFP; Homo sapiens; RNA-Seq     SRR648667      9606   \n",
       "1   GSM1063576: 293T_GFP_2hrs_severe_Heat_Shock; H...     SRR648668      9606   \n",
       "2      GSM1063577: 293T_Hspa1a; Homo sapiens; RNA-Seq     SRR648669      9606   \n",
       "3   GSM1063578: 293T_Hspa1a_2hrs_severe_Heat_Shock...     SRR648670      9606   \n",
       "4   GSM794854: 3T3-Control-Riboseq; Mus musculus; ...     SRR649752     10090   \n",
       "5   GSM794855: 3T3-8hrs-mild-HS-Riboseq; Mus muscu...     SRR649753     10090   \n",
       "6   GSM794856: 3T3-2hrs-severe-HS-Riboseq; Mus mus...     SRR649754     10090   \n",
       "7   GSM1063574: 3T3-Hsp70_inhibitor_3hr; Mus muscu...     SRR648666     10090   \n",
       "8   GSM794857: 3T3-Control-Riboseq-rep2; Mus muscu...     SRR649755     10090   \n",
       "9   GSM794858: 3T3-8hrs-mild-HS-Riboseq-rep2; Mus ...     SRR649756     10090   \n",
       "10  GSM794859: 3T3-2hrs-severe-HS-Riboseq-rep2; Mu...     SRR649757     10090   \n",
       "11  GSM794860: 3T3-8hrs-mild-2hrs-severe-HS-Ribose...     SRR649758     10090   \n",
       "12  GSM794848: 3T3-Control-RNAseq; Mus musculus; R...     SRR649749     10090   \n",
       "13  GSM794849: 3T3-8hrs-mild-HS-RNAseq; Mus muscul...     SRR649750     10090   \n",
       "14  GSM794850: 3T3-2hrs-severe-HS-RNAseq; Mus musc...     SRR649751     10090   \n",
       "\n",
       "   library_selection library_layout library_strategy  library_source  \\\n",
       "0              other      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "1              other      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "2              other      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "3              other      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "4               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "5               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "6               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "7              other      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "8               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "9               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "10              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "11              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "12              cDNA      PAIRED -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "13              cDNA      PAIRED -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "14              cDNA      PAIRED -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "\n",
       "   library_name       bases      spots adapter_spec  avg_read_length  \n",
       "0          None  1806641316   50184481         None             36.0  \n",
       "1          None  3436984836   95471801         None             36.0  \n",
       "2          None  3330909216   92525256         None             36.0  \n",
       "3          None  3622123512  100614542         None             36.0  \n",
       "4          None   594945396   16526261         None             36.0  \n",
       "5          None   730843272   20301202         None             36.0  \n",
       "6          None  1785538908   49598303         None             36.0  \n",
       "7          None  5932552920  148313823         None             40.0  \n",
       "8          None  3713863640   92846591         None             40.0  \n",
       "9          None  3967599960   99189999         None             40.0  \n",
       "10         None  3458211480   86455287         None             40.0  \n",
       "11         None  3891505080   97287627         None             40.0  \n",
       "12         None  1766712168   24537669         None             72.0  \n",
       "13         None  1960403976   27227833         None             72.0  \n",
       "14         None  1393136640   19349120         None             72.0  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = db.sra_convert('SRP017942')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>library_selection</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>library_strategy</th>\n",
       "      <th>library_source</th>\n",
       "      <th>library_name</th>\n",
       "      <th>bases</th>\n",
       "      <th>spots</th>\n",
       "      <th>adapter_spec</th>\n",
       "      <th>avg_read_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021966</td>\n",
       "      <td>GSM546920: Footprint_mock_32hr_runs1-2</td>\n",
       "      <td>SRR057511</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546920: Footprint_mock_32hr_runs1-2</td>\n",
       "      <td>173069316</td>\n",
       "      <td>4807481</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021966</td>\n",
       "      <td>GSM546920: Footprint_mock_32hr_runs1-2</td>\n",
       "      <td>SRR057512</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546920: Footprint_mock_32hr_runs1-2</td>\n",
       "      <td>670274352</td>\n",
       "      <td>18618732</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021967</td>\n",
       "      <td>GSM546921: mRNASeq_mock_32hr_runs1-3</td>\n",
       "      <td>SRR057513</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546921: mRNASeq_mock_32hr_runs1-3</td>\n",
       "      <td>159276276</td>\n",
       "      <td>4424341</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021967</td>\n",
       "      <td>GSM546921: mRNASeq_mock_32hr_runs1-3</td>\n",
       "      <td>SRR057515</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546921: mRNASeq_mock_32hr_runs1-3</td>\n",
       "      <td>92929104</td>\n",
       "      <td>2581364</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021968</td>\n",
       "      <td>GSM546922: Footprint_miR-155_32hr_runs1-2</td>\n",
       "      <td>SRR057516</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546922: Footprint_miR-155_32hr_runs1-2</td>\n",
       "      <td>212446440</td>\n",
       "      <td>5901290</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021968</td>\n",
       "      <td>GSM546922: Footprint_miR-155_32hr_runs1-2</td>\n",
       "      <td>SRR057517</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546922: Footprint_miR-155_32hr_runs1-2</td>\n",
       "      <td>522725688</td>\n",
       "      <td>14520158</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021969</td>\n",
       "      <td>GSM546923: mRNASeq_miR-155_32hr_runs1-3</td>\n",
       "      <td>SRR057518</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546923: mRNASeq_miR-155_32hr_runs1-3</td>\n",
       "      <td>216374004</td>\n",
       "      <td>6010389</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021969</td>\n",
       "      <td>GSM546923: mRNASeq_miR-155_32hr_runs1-3</td>\n",
       "      <td>SRR057519</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546923: mRNASeq_miR-155_32hr_runs1-3</td>\n",
       "      <td>207243972</td>\n",
       "      <td>5756777</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021969</td>\n",
       "      <td>GSM546923: mRNASeq_miR-155_32hr_runs1-3</td>\n",
       "      <td>SRR057520</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546923: mRNASeq_miR-155_32hr_runs1-3</td>\n",
       "      <td>94523436</td>\n",
       "      <td>2625651</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021970</td>\n",
       "      <td>GSM546924: Footprint_miR-1_32hr_runs1-2</td>\n",
       "      <td>SRR057521</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546924: Footprint_miR-1_32hr_runs1-2</td>\n",
       "      <td>212789088</td>\n",
       "      <td>5910808</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021970</td>\n",
       "      <td>GSM546924: Footprint_miR-1_32hr_runs1-2</td>\n",
       "      <td>SRR057522</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546924: Footprint_miR-1_32hr_runs1-2</td>\n",
       "      <td>510414552</td>\n",
       "      <td>14178182</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021971</td>\n",
       "      <td>GSM546925: mRNASeq_miR-1_32hr_runs1-3</td>\n",
       "      <td>SRR057524</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546925: mRNASeq_miR-1_32hr_runs1-3</td>\n",
       "      <td>204962256</td>\n",
       "      <td>5693396</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021971</td>\n",
       "      <td>GSM546925: mRNASeq_miR-1_32hr_runs1-3</td>\n",
       "      <td>SRR057525</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546925: mRNASeq_miR-1_32hr_runs1-3</td>\n",
       "      <td>89765496</td>\n",
       "      <td>2493486</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021972</td>\n",
       "      <td>GSM546926: Footprint_mock_12hr_run1</td>\n",
       "      <td>SRR057526</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546926: Footprint_mock_12hr_run1</td>\n",
       "      <td>733189464</td>\n",
       "      <td>20366374</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021973</td>\n",
       "      <td>GSM546927: mRNASeq_mock_12hr_runs1-2</td>\n",
       "      <td>SRR057527</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546927: mRNASeq_mock_12hr_runs1-2</td>\n",
       "      <td>966335760</td>\n",
       "      <td>26842660</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021973</td>\n",
       "      <td>GSM546927: mRNASeq_mock_12hr_runs1-2</td>\n",
       "      <td>SRR057528</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546927: mRNASeq_mock_12hr_runs1-2</td>\n",
       "      <td>68137020</td>\n",
       "      <td>1892695</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021974</td>\n",
       "      <td>GSM546928: Footprint_miR-155_12hr_run1</td>\n",
       "      <td>SRR057529</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546928: Footprint_miR-155_12hr_run1</td>\n",
       "      <td>728406504</td>\n",
       "      <td>20233514</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021975</td>\n",
       "      <td>GSM546929: mRNASeq_miR-155_12hr_runs1-2</td>\n",
       "      <td>SRR057530</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546929: mRNASeq_miR-155_12hr_runs1-2</td>\n",
       "      <td>979016184</td>\n",
       "      <td>27194894</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021975</td>\n",
       "      <td>GSM546929: mRNASeq_miR-155_12hr_runs1-2</td>\n",
       "      <td>SRR057531</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546929: mRNASeq_miR-155_12hr_runs1-2</td>\n",
       "      <td>89522964</td>\n",
       "      <td>2486749</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021976</td>\n",
       "      <td>GSM546930: Footprint_miR-1_12hr_run1</td>\n",
       "      <td>SRR057532</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546930: Footprint_miR-1_12hr_run1</td>\n",
       "      <td>677440620</td>\n",
       "      <td>18817795</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021977</td>\n",
       "      <td>GSM546931: mRNASeq_miR-1_12hr_runs1-2</td>\n",
       "      <td>SRR057533</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546931: mRNASeq_miR-1_12hr_runs1-2</td>\n",
       "      <td>921246408</td>\n",
       "      <td>25590178</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021977</td>\n",
       "      <td>GSM546931: mRNASeq_miR-1_12hr_runs1-2</td>\n",
       "      <td>SRR057534</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546931: mRNASeq_miR-1_12hr_runs1-2</td>\n",
       "      <td>87452964</td>\n",
       "      <td>2429249</td>\n",
       "      <td>None</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021967</td>\n",
       "      <td>GSM546921: mRNASeq_mock_32hr_runs1-3</td>\n",
       "      <td>SRR057514</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546921: mRNASeq_mock_32hr_runs1-3</td>\n",
       "      <td>239528394</td>\n",
       "      <td>4435711</td>\n",
       "      <td>None</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SRP002605</td>\n",
       "      <td>SRX021971</td>\n",
       "      <td>GSM546925: mRNASeq_miR-1_32hr_runs1-3</td>\n",
       "      <td>SRR057523</td>\n",
       "      <td>9606</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>SINGLE -</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>GSM546925: mRNASeq_miR-1_32hr_runs1-3</td>\n",
       "      <td>287181396</td>\n",
       "      <td>5318174</td>\n",
       "      <td>None</td>\n",
       "      <td>54.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   study_accession experiment_accession  \\\n",
       "0        SRP002605            SRX021966   \n",
       "1        SRP002605            SRX021966   \n",
       "2        SRP002605            SRX021967   \n",
       "3        SRP002605            SRX021967   \n",
       "4        SRP002605            SRX021968   \n",
       "5        SRP002605            SRX021968   \n",
       "6        SRP002605            SRX021969   \n",
       "7        SRP002605            SRX021969   \n",
       "8        SRP002605            SRX021969   \n",
       "9        SRP002605            SRX021970   \n",
       "10       SRP002605            SRX021970   \n",
       "11       SRP002605            SRX021971   \n",
       "12       SRP002605            SRX021971   \n",
       "13       SRP002605            SRX021972   \n",
       "14       SRP002605            SRX021973   \n",
       "15       SRP002605            SRX021973   \n",
       "16       SRP002605            SRX021974   \n",
       "17       SRP002605            SRX021975   \n",
       "18       SRP002605            SRX021975   \n",
       "19       SRP002605            SRX021976   \n",
       "20       SRP002605            SRX021977   \n",
       "21       SRP002605            SRX021977   \n",
       "22       SRP002605            SRX021967   \n",
       "23       SRP002605            SRX021971   \n",
       "\n",
       "                             experiment_title run_accession  taxon_id  \\\n",
       "0      GSM546920: Footprint_mock_32hr_runs1-2     SRR057511      9606   \n",
       "1      GSM546920: Footprint_mock_32hr_runs1-2     SRR057512      9606   \n",
       "2        GSM546921: mRNASeq_mock_32hr_runs1-3     SRR057513      9606   \n",
       "3        GSM546921: mRNASeq_mock_32hr_runs1-3     SRR057515      9606   \n",
       "4   GSM546922: Footprint_miR-155_32hr_runs1-2     SRR057516      9606   \n",
       "5   GSM546922: Footprint_miR-155_32hr_runs1-2     SRR057517      9606   \n",
       "6     GSM546923: mRNASeq_miR-155_32hr_runs1-3     SRR057518      9606   \n",
       "7     GSM546923: mRNASeq_miR-155_32hr_runs1-3     SRR057519      9606   \n",
       "8     GSM546923: mRNASeq_miR-155_32hr_runs1-3     SRR057520      9606   \n",
       "9     GSM546924: Footprint_miR-1_32hr_runs1-2     SRR057521      9606   \n",
       "10    GSM546924: Footprint_miR-1_32hr_runs1-2     SRR057522      9606   \n",
       "11      GSM546925: mRNASeq_miR-1_32hr_runs1-3     SRR057524      9606   \n",
       "12      GSM546925: mRNASeq_miR-1_32hr_runs1-3     SRR057525      9606   \n",
       "13        GSM546926: Footprint_mock_12hr_run1     SRR057526      9606   \n",
       "14       GSM546927: mRNASeq_mock_12hr_runs1-2     SRR057527      9606   \n",
       "15       GSM546927: mRNASeq_mock_12hr_runs1-2     SRR057528      9606   \n",
       "16     GSM546928: Footprint_miR-155_12hr_run1     SRR057529      9606   \n",
       "17    GSM546929: mRNASeq_miR-155_12hr_runs1-2     SRR057530      9606   \n",
       "18    GSM546929: mRNASeq_miR-155_12hr_runs1-2     SRR057531      9606   \n",
       "19       GSM546930: Footprint_miR-1_12hr_run1     SRR057532      9606   \n",
       "20      GSM546931: mRNASeq_miR-1_12hr_runs1-2     SRR057533      9606   \n",
       "21      GSM546931: mRNASeq_miR-1_12hr_runs1-2     SRR057534      9606   \n",
       "22       GSM546921: mRNASeq_mock_32hr_runs1-3     SRR057514      9606   \n",
       "23      GSM546925: mRNASeq_miR-1_32hr_runs1-3     SRR057523      9606   \n",
       "\n",
       "   library_selection library_layout library_strategy  library_source  \\\n",
       "0               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "1               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "2               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "3               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "4               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "5               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "6               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "7               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "8               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "9               cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "10              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "11              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "12              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "13              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "14              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "15              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "16              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "17              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "18              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "19              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "20              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "21              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "22              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "23              cDNA      SINGLE -           RNA-Seq  TRANSCRIPTOMIC   \n",
       "\n",
       "                                 library_name      bases     spots  \\\n",
       "0      GSM546920: Footprint_mock_32hr_runs1-2  173069316   4807481   \n",
       "1      GSM546920: Footprint_mock_32hr_runs1-2  670274352  18618732   \n",
       "2        GSM546921: mRNASeq_mock_32hr_runs1-3  159276276   4424341   \n",
       "3        GSM546921: mRNASeq_mock_32hr_runs1-3   92929104   2581364   \n",
       "4   GSM546922: Footprint_miR-155_32hr_runs1-2  212446440   5901290   \n",
       "5   GSM546922: Footprint_miR-155_32hr_runs1-2  522725688  14520158   \n",
       "6     GSM546923: mRNASeq_miR-155_32hr_runs1-3  216374004   6010389   \n",
       "7     GSM546923: mRNASeq_miR-155_32hr_runs1-3  207243972   5756777   \n",
       "8     GSM546923: mRNASeq_miR-155_32hr_runs1-3   94523436   2625651   \n",
       "9     GSM546924: Footprint_miR-1_32hr_runs1-2  212789088   5910808   \n",
       "10    GSM546924: Footprint_miR-1_32hr_runs1-2  510414552  14178182   \n",
       "11      GSM546925: mRNASeq_miR-1_32hr_runs1-3  204962256   5693396   \n",
       "12      GSM546925: mRNASeq_miR-1_32hr_runs1-3   89765496   2493486   \n",
       "13        GSM546926: Footprint_mock_12hr_run1  733189464  20366374   \n",
       "14       GSM546927: mRNASeq_mock_12hr_runs1-2  966335760  26842660   \n",
       "15       GSM546927: mRNASeq_mock_12hr_runs1-2   68137020   1892695   \n",
       "16     GSM546928: Footprint_miR-155_12hr_run1  728406504  20233514   \n",
       "17    GSM546929: mRNASeq_miR-155_12hr_runs1-2  979016184  27194894   \n",
       "18    GSM546929: mRNASeq_miR-155_12hr_runs1-2   89522964   2486749   \n",
       "19       GSM546930: Footprint_miR-1_12hr_run1  677440620  18817795   \n",
       "20      GSM546931: mRNASeq_miR-1_12hr_runs1-2  921246408  25590178   \n",
       "21      GSM546931: mRNASeq_miR-1_12hr_runs1-2   87452964   2429249   \n",
       "22       GSM546921: mRNASeq_mock_32hr_runs1-3  239528394   4435711   \n",
       "23      GSM546925: mRNASeq_miR-1_32hr_runs1-3  287181396   5318174   \n",
       "\n",
       "   adapter_spec  avg_read_length  \n",
       "0          None             36.0  \n",
       "1          None             36.0  \n",
       "2          None             36.0  \n",
       "3          None             36.0  \n",
       "4          None             36.0  \n",
       "5          None             36.0  \n",
       "6          None             36.0  \n",
       "7          None             36.0  \n",
       "8          None             36.0  \n",
       "9          None             36.0  \n",
       "10         None             36.0  \n",
       "11         None             36.0  \n",
       "12         None             36.0  \n",
       "13         None             36.0  \n",
       "14         None             36.0  \n",
       "15         None             36.0  \n",
       "16         None             36.0  \n",
       "17         None             36.0  \n",
       "18         None             36.0  \n",
       "19         None             36.0  \n",
       "20         None             36.0  \n",
       "21         None             36.0  \n",
       "22         None             54.0  \n",
       "23         None             54.0  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = db.sra_convert('SRP002605')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "b''\n",
      "b'Session Stop  (Error: Client unable to connect to server (check UDP port and firewall))'\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-a8de7243afe6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msra_convert\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'SRP098789'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdownload\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mascp_bin\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'/usr/bin/ascp'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pysradb-0.1.0-py3.6.egg/pysradb/pysradb.py\u001b[0m in \u001b[0;36mdownload\u001b[0;34m(self, srp, df, out_dir, protocol, ascp_bin)\u001b[0m\n\u001b[1;32m    231\u001b[0m             cmd = '{} {} {} {}'.format(cmd, _find_aspera_keypath(), url,\n\u001b[1;32m    232\u001b[0m                                        srx_dir)\n\u001b[0;32m--> 233\u001b[0;31m             \u001b[0mrun_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    234\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/pysradb-0.1.0-py3.6.egg/pysradb/pysradb.py\u001b[0m in \u001b[0;36mrun_command\u001b[0;34m(command)\u001b[0m\n\u001b[1;32m     51\u001b[0m         shlex.split(command), stdout=subprocess.PIPE, stderr=subprocess.STDOUT)\n\u001b[1;32m     52\u001b[0m     \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m         \u001b[0moutput\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     54\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0moutput\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m''\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpoll\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m             \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "df = db.sra_convert('SRP098789')\n",
    "db.download(df=df, ascp_bin='/usr/bin/ascp')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
